articles/implementing-firebase-in-flutter/index.md

```diff
@@ -0,0 +1,258 @@
+
+```json
+{
+  "rules": {
+    ".read": "now < 1610053200000",  // 2021-1-8
+    ".write": "now < 1610053200000",  // 2021-1-8
+  }
+}
+```
+
+We have successfully set up Firebase and the flutter project. Let's design the UI.
+
+### Creating UI
+The app will have a simple user interface. It will allow a user to input a word or sentence and click on a button to save it to the Firebase database.
+
+![ui](/engineering-educationimplementing-firebase-in-flutter/ui.jpeg)
+
+Create a new file called `home.dart.` In this file, create a `stateful class` and name it `HomePage.`
+
+We can then create the above `UI` in this file by pasting the following code in the body section of the `Scaffold.`
+
+```dart
+Container(
+  child: Column(
+    children: <Widget>[
+      SizedBox(height: 250.0),
+      Padding(
+        padding: EdgeInsets.all(10.0),
+        child: TextField(
+        ),
+      ),
+      SizedBox(height: 30.0),
+      Center(
+          child: RaisedButton(
+              color: Colors.pinkAccent,
+              child: Text("Save to Database"),
+              onPressed: () {
+                addData(textcontroller.text);
+                //call method flutter upload
+              })),
+    ],
+  ),
+);
+```
+
+Import the `HomePage` class in the `main.dart` file. Ensure that the `HomePage` is initialized, as shown below.
+
+```dart
+import 'package:flutter/material.dart';
+import 'home.dart';
+import 'package:firebase_core/firebase_core.dart';
+
+
+void main(){
+  runApp(MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  // This widget is the root of your application.
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+        visualDensity: VisualDensity.adaptivePlatformDensity,
+      ),
+      home: HomePage(),
+    );
+  }
+}
+```
+
+### 3. Storing data
+We will initialize a Firebase `databaseReference` object and use it to store data.
+
+```dart
+final databaseRef = FirebaseDatabase._instance_.reference(); //database reference object
+
+void addData(String data) {
+  databaseRef.push().set({'name': data, 'comment': 'A good season'});
+}
+```
+
+Please note that we need to use a `FutureBuilder` for long-running operations. This class allows the application to retrieve results once the network operation is completed. This class also helps us initialize the Firebase app. You can learn more about FutureBuilder from [here](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html).
+
+### 4. Retrieving data
+For simplicity, we will retrieve and print out the data in the console. This is shown below.
+
+```dart
+void printFirebase(){
+  databaseRef.once().then((DataSnapshot snapshot) {
+    print('Data : ${snapshot.value}');
+  });
+}
+```
```

Comment:
Can you mention where to add this snippet and when this function will get called?